code/game/machinery/atmoalter/canister.dm

```diff
@@ -62,33 +80,76 @@
 	can_label = 0
 
 /obj/machinery/portable_atmospherics/canister/update_icon()
-	overlays = null
-
-	if (destroyed)
+	if(destroyed)
 		icon_state = "[canister_color]-1"
-	else
-		icon_state = canister_color
-		overlays = new/list()
+		return
 
-		if (holding)
-			overlays.Add("can-open")
+	if(!status_overlays)
+		status_overlays = 1
 
-		if (connected_port)
-			overlays.Add("can-connector")
+		status_overlays_pressure = new
+		status_overlays_other = new
 
-		var/tank_pressure = air_contents.return_pressure()
+		status_overlays_pressure.len = 4
+		status_overlays_other.len = 2
 
-		if (tank_pressure < 10)
-			overlays.Add("can-o0")
-		else if (tank_pressure < ONE_ATMOSPHERE)
-			overlays.Add("can-o1")
-		else if (tank_pressure < 15 * ONE_ATMOSPHERE)
-			overlays.Add("can-o2")
-		else
-			overlays.Add("can-o3")
+		status_overlays_pressure[1] = image(icon, "can-o0")
+		status_overlays_pressure[2] = image(icon, "can-o1")
+		status_overlays_pressure[3] = image(icon, "can-o2")
+		status_overlays_pressure[4] = image(icon, "can-o3")
+
+		status_overlays_other[1]  = image(icon, "can-open")
+		status_overlays_other[2]  = image(icon, "can-connector")
+
+	if (canister_color != old_color)
+		icon_state = "[canister_color]-1"
+		old_color = canister_color
+
+	var/tank_pressure = air_contents.return_pressure()
+	if(check_updates(tank_pressure))
+		if(overlays.len)
+			overlays = 0
 
+		overlay_status = 0
+
+		if (holding)
+			overlays += status_overlays_other[1]
+			overlay_status |= OVERLAY_HOLDING
+
+		if (connected_port)
+			overlays += status_overlays_other[2]
+			overlay_status |= OVERLAY_CONNECTED
+
+		switch(tank_pressure)
+			if(15 * ONE_ATMOSPHERE to INFINITY)

```

Comment:
No, they're plucked from the air just like tanks holding 10 atmospheres and being filled with 3 or 6